### DISCUSSION TITLE
How do you document your Gateway endpoints?

### DISCUSSION THREAD

**Discussion by NadChel** [community/community#72609] [Programming Help]
### Body

Do you think you can give me some advice on how I provide documentation for all my Gateway endpoints? You can't annotate this as a regular synchronous REST controller:
```java
    @Bean
    public RouteLocator creditServiceRoutes(RouteLocatorBuilder builder) {
        return builder.routes()
                .route(p -> p
                        .method(CC1.method()) // CC1, CC2, etc. refers to how the endpoints are documented
                        .and()
                        .path(CC1.gatewayUri())
                        .filters(f -> CommonConfig.replaceWithAuthSpec(f)
                                        .filter(addClientIdParamFilter))
                        .uri(CREDIT_SERVICE_URL))
                .route(p -> p
                        .method(CC2.method())
                        .and()
                        .path(CC2.gatewayUri())
                        .and()
                        .query("cardId")
                        .filters(f -> CommonConfig.replaceWithAuthSpec(f)
                                        .filter(addClientIdParamFilter))
                        .uri(CREDIT_SERVICE_URL))
```
There's a theory that you can:

1. Provide Swagger dependencies in services
2. Annotate handler methods
3. Swagger would generate a documentation JSON and provide an endpoint for retrieving it
4. Your Gateway can (on startup!) go to Eureka, retrieve all the endpoints
5. Then visit all those doc endpoints (`credit/swagger-config.json`, `money-transfer/swagger-config.json`, etc.)
6. The locator bean waits till it's all done and then you can refer to those endpoints in you locator bean method

It may be a little patchy. Can you give me a few suggestions?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by arienshibani** [comment]
Here are some recommendations for providing documentation for all your Gateway endpoints:

- **Inline Comments:** Add inline comments to each section of the code to explain the purpose of different configurations, such as the purpose of CC1 and CC2, and what each section of the routing configuration does. Provide clear and concise explanations to make the code more understandable.

- **External Documentation:** Create external documentation, preferably in Markdown format, to explain the purpose of the Gateway endpoints, the filters applied, and the routing configurations used. You can include this documentation in the project's README file or in a separate documentation file within your project's repository.

- **API Documentation Tool:** Utilize API documentation tools like Swagger or Springdoc to generate API documentation for your Gateway endpoints automatically. These tools can help you document the routes, methods, paths, and filters used in your Gateway configuration. You can integrate the generated documentation with your project to provide comprehensive information about your Gateway endpoints.

- **Use Descriptive Naming:** Ensure that you use descriptive and meaningful names for your variables, methods, and filters. This makes it easier for other developers to understand the purpose of each component in your Gateway configuration without additional documentation.

**Comment by NadChel** [comment]
> Here are some recommendations for providing documentation for all your Gateway endpoints:
> 
> * **Inline Comments:** Add inline comments to each section of the code to explain the purpose of different configurations, such as the purpose of CC1 and CC2, and what each section of the routing configuration does. Provide clear and concise explanations to make the code more understandable.
> * **External Documentation:** Create external documentation, preferably in Markdown format, to explain the purpose of the Gateway endpoints, the filters applied, and the routing configurations used. You can include this documentation in the project's README file or in a separate documentation file within your project's repository.
> * **API Documentation Tool:** Utilize API documentation tools like Swagger or Springdoc to generate API documentation for your Gateway endpoints automatically. These tools can help you document the routes, methods, paths, and filters used in your Gateway configuration. You can integrate the generated documentation with your project to provide comprehensive information about your Gateway endpoints.
> * **Use Descriptive Naming:** Ensure that you use descriptive and meaningful names for your variables, methods, and filters. This makes it easier for other developers to understand the purpose of each component in your Gateway configuration without additional documentation.

Hi! Thank you for your feedback

Here's a specific aspect I am wondering about. Here's a snippet from an on old-fashioned locator bean with hard-coded values:
```java
    @Bean
    public RouteLocator creditServiceRoutes(RouteLocatorBuilder builder) {
        return builder.routes()

                .route(p -> p
                        .path("/api/v1/credit-cards")
                        .filters(f -> f.filter(new AddClientIdParamToRequest())
                                .rewritePath("/api/v1", "/auth"))
                        .uri(CREDIT_SERVICE_URL))
```
How do I modify it to perform routing in a new and sexy way that involves reading from dynamically fetched Swagger docs that I fetch like this (see the snippet below) while performing the rewriting from the "old" locator bean? For example if a Gateway receives something like `/api/v1/example`, I want it to route it to `/auth/example` (provided `/auth/example` is one of the endpoints we got in `fetchDocs()`)
```java
// warning! it's a rough approximation
    private void fetchDocs() {
        List<Mono<String>> docs = new ArrayList<>();
        endpointDatas.forEach(service ->
                docs.add(
                        webClientBuilder
                                .build()
                                .get()
                                .uri(
                                        MessageFormat.format("{}{}", service.getServiceUrl(), API_DOCS_PATH)
                                )
                                /*
                                There are a few things you should know. service.getServiceUrl() returns things like "lb://SOME_SERVICE" (we use Eureka)
                                and API_DOCS_PATH is "/v3/api-docs" (a standard Swagger/OpenAPI endpoint)
                                endpointDatas is a list of objects of a custom type EndpointData that is basically a simple data class with two fields:
                                serviceName (like "SOME-SERVICE") and serviceUrl which is serviceName prepended 
                                with "lb://" (like "lb://SOME-SERVICE")
                                */
                                .retrieve()
                                .bodyToMono(String.class)
                )
        );
        Flux.merge(docs).subscribe(
        // something should happen here, I guess?
        );
        // we may save it to some field fetchable with a public getter?..
    }
```
Can you guide me a little?
